heartbeat/ocf-shellfuncs.in

```diff
@@ -672,6 +672,18 @@ EOF
 	systemctl daemon-reload
 }
 
+# move process to root cgroup if realtime scheduling is enabled
+ocf_move_to_root_cgroup_if_rt_enabled()
+{
+	if [ -e "/sys/fs/cgroup/cpu/cpu.rt_runtime_us" ]; then
```

Comment:
Has this been verified on RHEL 9? I haven't done anything intentional to configure realtime scheduling on either 8 or 9. However, on 8, `/sys/fs/cgroup/cpu` exists, while on 9, it does not.

Reply:
This isnt enabled on RHEL9, which is why we check if the realtime file exists before doing anything, so we dont break anything on RHEL9 or other distros not having it enabled.

Reply:
Is it not going to be needed for RHEL 9?

It looks like moving a process to the root cgroup on RHEL 9 (with cgroups v2) is as simple as
```
echo $$ >> /sys/fs/cgroup/cgroup.procs
```

https://man7.org/conf/lca2019/cgroups_v2-LCA2019-Kerrisk.pdf
Did basic testing and it seems plausible.
```
[root@pcmk-1 cgroup]# cat ~/test.sh
#!/bin/bash

cgroup_proc_files=$(find /sys/fs/cgroup -name cgroup.procs)
our_pid=$$
echo "our pid: $our_pid"
echo "before"
for file in $cgroup_proc_files; do
	grep $our_pid $file && echo "found in $file!"
done

echo $our_pid >> /sys/fs/cgroup/cgroup.procs

echo "after"
for file in $cgroup_proc_files; do
	grep $our_pid $file && echo "found in $file!"
done


[root@pcmk-1 cgroup]# bash ~/test.sh
our pid: 1034593
before
1034593
found in /sys/fs/cgroup/user.slice/user-0.slice/session-3.scope/cgroup.procs!
after
1034593
found in /sys/fs/cgroup/cgroup.procs!
```

Reply:
Just want to make sure we're on the same page that RHEL 9 uses a totally different cgroup hierarchy. It doesn't mean the cpu controller is not enabled. On the contrary, the cpu controller is enabled:
```
[root@pcmk-1 cgroup]# cat cgroup.controllers 
cpuset cpu io memory hugetlb pids rdma misc
```
but there won't be a `/sys/fs/cgroup/cpu` due to the switch to cgroups v2.

Reply:
Okay, the above `cgroup.controllers` might just mean the controller is available... still I think it can **become** enabled and `/sys/fs/cgroup/cpu` still won't exist.

Reply:
Yeah. I'll have to find the correct way to detect it there as well.

Reply:
and I think the realtime-budget-stuff isn't enabled on rhel 9 anymore.
moving to root-cgroup led to loosing all systemd-logging for the process iirc.
checking for the path as done in the ra won't just check if the realtime-budgeting
is enabled but as well it will just trigger on old cgroup config. for current rhel 9
config the decision should still be fine - for older rhel 9 kernels not - but that
is probably not interesting anyway.

Reply:
I think `grep -w cpu /sys/fs/cgroup/cgroup.subtree_control` tells whether it's enabled.

Reply:
@nrwahl2 Luckilly RT scheduling cgroup is NOT enabled in RHEL 9 (kernel config doesn't contain CONFIG_RT_GROUP_SCHED ). It is simply impossible to make it work with RHEL 9, because of cgroup v2. Been there (for corosync), and memories really hurt a LOT (doesn't work properly and breaks journald - see commit message https://github.com/corosync/corosync/commit/57e6b86b53010dd2612b0a6a4e04917673062ecf, https://github.com/corosync/corosync/commit/fc7b420e944c9c589966b5c73171da50e8ee809b  and finally https://github.com/corosync/corosync/commit/c9996fdd0f4fa1fbf113b740eea01bcc70b235aa)- and this was the reason why I've suggested to remove that kernel option (https://bugzilla.redhat.com/show_bug.cgi?id=1980362 and problems with journald https://bugzilla.redhat.com/show_bug.cgi?id=1962768 (suprise suprise, ignored by systemd developers)

Reply:
So to conclude, this patch (as it is) is perfectly fine. ACK by me

Reply:
Yep. Should be exactly what corosync is doing atm - right?
What could be done on top is checking if the move to root-cgroup has to be done even with the budgeting enabled.
If there is enough budget - for whatever reason - maybe because sbdy has introduced a realtime-slice anyway - we
wouldn't have to do it. Didn't check how complicated the check would have to be as the budget might be configurable for ctdbd.

Reply:
Also, there's no CTDB in RHEL9.

Reply:
> Also, there's no CTDB in RHEL9.

*applause*

Reply:
my last comment wasn't for rhel9.
even if there is no CTDB on distros using cgroup v2 setup ocf_move_to_root_cgroup_if_rt_enabled
is a generic utility-function we could leave in.
then it might make sense to pass it a budget to check for and it could issue a warning when
running on cgroup v2 that it is not able to check (I found a way but it is ugly) or move (we've seen
it to be problematic with systemd-logging at least).